lua/neo-tree/sources/common/commands.lua

```diff
@@ -343,6 +346,53 @@ M.delete_visual = function(state, selected_nodes, callback)
   fs_actions.delete_nodes(paths_to_delete, callback)
 end
 
+M.preview = function(state)
+  local node = state.tree:get_node()
+  if state.current_position == "current" or node.type == "directory" then
+    return
+  end
+
+  if not state.preview then
+    state.preview = Preview:new(state)
+  else
+    state.preview:findWindow(state)
+  end
+
+  local extra = node.extra or {}
+  local position = extra.position
+  local end_position = extra.end_position
+  local bufnr = extra.bufnr
+  if bufnr == nil then
+    local path = node.path or node:get_id()
+    pcall(vim.cmd, "badd " .. path)
+    bufnr = utils.find_buffer_by_name(path)
+  end
+
+  if bufnr and bufnr > 0 then
+    state.preview:preview(bufnr, position, end_position)
+  end
+end
+
+M.revert_preview = function(state)
+  if state.preview and state.preview.active then
+    state.preview:revert()
+  end
+end
+
+M.toggle_preview = function(state)
+  local preview_event = {
+    event = events.VIM_CURSOR_MOVED,
+    handler = utils.wrap(M.preview, state),
+    id = "preview-event",
+  }
+  if state.preview and state.preview.active then
+    state.preview:revert()
+    manager.unsubscribe(state.name, preview_event)
+  else
```

Comment:
What if the Neo-tree window is closed or the source is changed? I think it should probably end the preview mode. I think this could be handled by subscribing to `NeoTreeBufferLeave`, which would fire when the Neo-tree buffer loses focus for any reason.

Reply:
I was hesitant to add it, but you're probably right. My thinking was that it would not be good for sources which benefit from activating the preview by default (such as diagnostics). How do you think it would be best to handle that? An event handler in the source's setup function should be enough, right?

Reply:
> An event handler in the source's setup function should be enough, right?

Yes. I think if you want to activate it automatically when showing a certain source, then you can subscribe to `NeoTreeBufferEnter` or `NeoTreeWindowAfterOpen` in the source's setup function, depending on what you want. I would not want to automatically enter preview mode though, so I would make that functionality optional.